File: bindings/python/tests/test_batch.py
"""BatchPool encode-batch tests, including leak checks via weakref."""

from __future__ import annotations

import gc
import weakref

import pytest

import ztok


def test_batch_pool_workers_resolves_auto() -> None:
    with ztok.BatchPool(workers=0) as pool:
        assert pool.workers >= 1


def test_batch_pool_explicit_worker_count() -> None:
    with ztok.BatchPool(workers=3) as pool:
        assert pool.workers == 3


def test_batch_pool_closed_pool_raises() -> None:
    pool = ztok.BatchPool(workers=2)
    pool.close()
    with pytest.raises(ztok.ZtokError):
        _ = pool.workers


def test_batch_encode_matches_single_encode(bpe_pipeline: ztok.Pipeline) -> None:
    inputs = ["hello world", " the quick brown fox", "foo bar baz"]
    expected = [bpe_pipeline.encode(s) for s in inputs]
    with ztok.BatchPool(workers=4) as pool:
        got = bpe_pipeline.encode_batch(pool, inputs)
    assert got == expected


def test_batch_encode_1000_strings_no_leaks(bpe_pipeline: ztok.Pipeline) -> None:
    inputs = ["hello world"] * 1000

    class _Holder:
        # Plain `list` doesn't support weakref; wrap so we can prove
        # the results object itself is reclaimable after we drop it.
        __slots__ = ("data", "__weakref__")

        def __init__(self, data):  # type: ignore[no-untyped-def]
            self.data = data

    with ztok.BatchPool(workers=8) as pool:
        holder = _Holder(bpe_pipeline.encode_batch(pool, inputs))
    assert len(holder.data) == 1000
    assert all(r == holder.data[0] for r in holder.data)
    # Weakref-style leak guard: the per-id-buffer C allocations were
    # freed inside encode_batch (via ztok_ids_free). Python-side, the
    # only references are the plain int lists we returned, which the
    # garbage collector reclaims cleanly once we drop the holder.
    ref = weakref.ref(holder)
    del holder
    gc.collect()
    assert ref() is None


def test_batch_encode_empty_inputs(bpe_pipeline: ztok.Pipeline) -> None:
    with ztok.BatchPool(workers=2) as pool:
        assert bpe_pipeline.encode_batch(pool, []) == []


def test_batch_encode_with_empty_string(bpe_pipeline: ztok.Pipeline) -> None:
    inputs = ["hello", "", "world"]
    with ztok.BatchPool(workers=2) as pool:
        results = bpe_pipeline.encode_batch(pool, inputs)
    assert len(results) == 3
    assert results[1] == []  # empty input -> empty ids
    assert results[0] == bpe_pipeline.encode("hello")
    assert results[2] == bpe_pipeline.encode("world")


def test_pool_finalizer_releases_on_gc() -> None:
    # Drop the pool without explicit close; the weakref.finalize hook
    # should still tear down the C worker pool when Python collects it.
    pool = ztok.BatchPool(workers=2)
    finalizer_alive = pool._finalizer.alive
    assert finalizer_alive
    ref = weakref.ref(pool)
    del pool
    gc.collect()
    assert ref() is None

File: bindings/ruby/test/test_batch.rb
# frozen_string_literal: true

require_relative "helper"

class TestBatch < Minitest::Test
  include ZtokTestFixtures

  def setup
    @pipe = Ztok::Pipeline.from_tiktoken(ZtokTestFixtures.tiktoken_fixture, cl100k: true)
  end

  def teardown
    @pipe&.close
  end

  def test_workers_resolves_auto
    pool = Ztok::BatchPool.new(workers: 0)
    begin
      assert_operator pool.workers, :>=, 1
    ensure
      pool.close
    end
  end

  def test_explicit_worker_count
    pool = Ztok::BatchPool.new(workers: 3)
    begin
      assert_equal 3, pool.workers
    ensure
      pool.close
    end
  end

  def test_closed_pool_raises
    pool = Ztok::BatchPool.new(workers: 2)
    pool.close
    assert pool.closed?
    assert_raises(Ztok::Error) { pool.workers }
  end

  def test_batch_encode_matches_single_encode
    inputs = ["hello world", " the quick brown fox", "foo bar baz"]
    expected = inputs.map { |s| @pipe.encode(s) }
    Ztok::BatchPool.open(workers: 4) do |pool|
      got = @pipe.encode_batch(pool, inputs)
      assert_equal expected, got
    end
  end

  def test_batch_encode_1000_strings_no_leaks
    inputs = Array.new(1000, "hello world")
    results = nil
    Ztok::BatchPool.open(workers: 8) do |pool|
      results = @pipe.encode_batch(pool, inputs)
    end
    assert_equal 1000, results.length
    first = results.first
    assert(results.all? { |r| r == first })

    # The C-owned id buffers were freed inside encode_batch (via
    # ztok_ids_free). Force a couple of GC cycles to flush our Ruby
    # finalizers and confirm we don't crash or leak.
    results = nil
    3.times { GC.start }
  end

  def test_batch_encode_empty_inputs
    Ztok::BatchPool.open(workers: 2) do |pool|
      assert_equal [], @pipe.encode_batch(pool, [])
    end
  end

  def test_batch_encode_with_empty_string
    inputs = ["hello", "", "world"]
    Ztok::BatchPool.open(workers: 2) do |pool|
      results = @pipe.encode_batch(pool, inputs)
      assert_equal 3, results.length
      assert_equal [], results[1] # empty input -> empty ids
      assert_equal @pipe.encode("hello"), results[0]
      assert_equal @pipe.encode("world"), results[2]
    end
  end

  def test_block_form_closes_pool
    captured = nil
    Ztok::BatchPool.open(workers: 2) do |pool|
      captured = pool
      assert_operator pool.workers, :>=, 1
    end
    assert captured.closed?, "block-form must close pool on exit"
  end

  def test_finalizer_releases_pool_on_gc
    # Drop without explicit close. The ObjectSpace finalizer should
    # tear down the C worker pool when Ruby collects the wrapper. We
    # don't have a portable "did the finalizer fire?" probe in MRI, but
    # we exercise the path so leaks surface as crashes under valgrind.
    _pool = Ztok::BatchPool.new(workers: 2)
    _pool = nil
    3.times { GC.start }
  end
end
